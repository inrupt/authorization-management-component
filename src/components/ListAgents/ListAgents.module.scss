@use "../../../scss/_variables" as *;

@import "../../../node_modules/bootstrap/scss/mixins";

.empty-inbox {
  font-size: 3em;
}

.empty-grant-container {
  text-align: center;
  font-size: 1.2em;
}

.empty-grant-container strong {
  font-weight: 900;
  font-size: 1.1em;
}
