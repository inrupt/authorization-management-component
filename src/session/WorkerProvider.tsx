//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
import {
  ReactElement,
  ReactNode,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

import { v4 } from "uuid";
import { isValidAccessGrant as _isValidAccessGrant } from "@inrupt/solid-client-access-grants";
import {
  getResourceInfo,
  getContentType as solidClientGetContentType,
} from "@inrupt/solid-client";
import { PurposeInfo, PurposeRequest, PurposeResponse } from "./workertypes";
import { SessionContext } from "./SessionProvider";
import {
  getFromWebIdHelper,
  namePredicates,
  imagePredicates,
} from "../helpers/profile/profile";

export interface IWorkerContext {
  getPurposeInfo: (_url: string) => Promise<PurposeInfo>;
  getNameFromWebId: (_url: string) => Promise<string | null>;
  getImageFromWebId: (_url: string) => Promise<string | null>;
  getContentType: (_url: string) => Promise<string | null>;
  isValidAccessGrant: (
    ..._args: Parameters<typeof _isValidAccessGrant>
  ) => Promise<boolean>;
  nameRecord: Record<string, string>;
}

export const WorkerContext = createContext<IWorkerContext>({
  getPurposeInfo: () =>
    new Promise<PurposeInfo>(() => {
      /* promise never resolves */
    }),
  getNameFromWebId: () =>
    new Promise<string | null>(() => {
      /* promise never resolves */
    }),
  getImageFromWebId: () =>
    new Promise<string | null>(() => {
      /* promise never resolves */
    }),
  getContentType: () =>
    new Promise<string | null>(() => {
      /* promise never resolves */
    }),
  isValidAccessGrant: async (vc, options) => {
    const valid = await _isValidAccessGrant(vc, options);
    return valid.errors.length === 0;
  },
  nameRecord: {},
});

/* eslint react/require-default-props: 0 */
export interface IWorkerProvider {
  children: ReactNode;
}

/**
 * Used to provide session data to child components through context.
 */
export default function WorkerProvider({
  children,
}: IWorkerProvider): ReactElement {
  const { fetch } = useContext(SessionContext);
  const [worker, setWorker] = useState<Worker>();
  const nameRecord = useMemo<Record<string, Promise<string | null>>>(
    () => ({}),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [fetch]
  );
  const imageRecord = useMemo<Record<string, Promise<string | null>>>(
    () => ({}),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [fetch]
  );
  const purposeRecord = useMemo<Record<string, PurposeInfo>>(() => ({}), []);
  const invalidatedUrl = useMemo<Set<string>>(() => new Set(), []);
  const [resolvedNameRecord, setResolvedNameRecord] = useState<
    Record<string, string>
  >({});

  useEffect(() => {
    setWorker(new Worker(new URL("./worker.ts", import.meta.url)));
  }, []);

  const getImageFromWebId = useCallback(
    // eslint-disable-next-line no-return-assign
    async (url: string) =>
      (imageRecord[url] ??= getFromWebIdHelper(url, imagePredicates, {
        fetch,
        allowedNamed: true,
      })),
    [fetch, imageRecord]
  );

  const getNameFromWebId = useCallback(
    // eslint-disable-next-line no-return-assign
    async (url: string) =>
      (nameRecord[url] ??= getFromWebIdHelper(url, namePredicates, {
        fetch,
      })).then((name) => {
        if (name && !(name in resolvedNameRecord)) {
          setResolvedNameRecord({ ...resolvedNameRecord, [url]: name });
        }
        return name;
      }),
    // We don't want to make resolvedNameRecord a dependency of the callback
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [fetch, nameRecord]
  );

  const getContentType = useMemo(() => {
    const cache: Record<string, Promise<string | null>> = {};
    return async (resourceIri: string) => {
      if (!(resourceIri in cache)) {
        cache[resourceIri] = getResourceInfo(resourceIri, {
          fetch,
        })
          .then((c) => solidClientGetContentType(c))
          .catch(() => {
            delete cache[resourceIri];
            return null;
          });
      }
      return cache[resourceIri];
    };
  }, [fetch]);

  const isValidAccessGrant = useCallback(
    async (...args: Parameters<typeof _isValidAccessGrant>) => {
      let [vcUrl] = args;

      if (
        typeof vcUrl === "object" &&
        !(vcUrl instanceof URL) &&
        "id" in vcUrl
      ) {
        vcUrl = vcUrl.id;
      }

      if (invalidatedUrl.has(vcUrl.toString())) {
        return false;
      }

      const result = await _isValidAccessGrant(...args);
      if (result.errors.length !== 0) {
        invalidatedUrl.add(vcUrl.toString());
        return false;
      }
      return true;
    },
    // We don't want to make resolvedNameRecord a dependency of the callback
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  const getPurposeInfo = useCallback(
    async (url: string) => {
      if (url in purposeRecord) {
        return purposeRecord[url];
      }

      return new Promise<PurposeInfo>((res) => {
        const fixedWorker = worker;
        const id = v4();

        const listener = ({ data }: MessageEvent) => {
          const result: PurposeResponse = JSON.parse(data);
          if (result.id === id) {
            fixedWorker?.removeEventListener("message", listener);
            purposeRecord[url] = result.value;
            res(result.value);
          }
        };
        fixedWorker?.addEventListener("message", listener);
        fixedWorker?.postMessage(
          JSON.stringify({
            type: "purpose",
            value: url,
            id,
          } as PurposeRequest)
        );
      });
    },
    [worker, purposeRecord]
  );

  const value = useMemo(
    () => ({
      getPurposeInfo,
      getNameFromWebId,
      getImageFromWebId,
      isValidAccessGrant,
      nameRecord: resolvedNameRecord,
      getContentType,
    }),
    [
      getPurposeInfo,
      getNameFromWebId,
      getImageFromWebId,
      resolvedNameRecord,
      isValidAccessGrant,
      getContentType,
    ]
  );

  return (
    <WorkerContext.Provider value={value}>{children}</WorkerContext.Provider>
  );
}
