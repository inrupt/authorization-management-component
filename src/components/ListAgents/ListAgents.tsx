// MIT License
//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import { useContext, useMemo } from "react";
import { Table } from "reactstrap";
import { comp, useOrdered } from "../../hooks/useOrdered";
import { WorkerContext } from "../../session/WorkerProvider";
import { DisplayAgent, DisplayImage } from "../PromiseText/PromiseText";
import { DetailsIcon, Thead } from "../SortedTable/SortedTable";
import styles from "../SortedTable/SortedTable.module.scss";
import style from "./ListAgents.module.scss";

const headings = [
  {
    key: "image",
    name: "Icon",
    hidden: true,
  },
  {
    key: "name",
    name: "Name",
    sortBy: comp,
  },
  {
    key: "webId",
    name: "WebID",
    sortBy: comp,
  },
  {
    key: "key",
    name: "Resources",
  },
];

function EmptyAgentPage() {
  return (
    <div className={`${style["empty-grant-container"]}`}>
      <i className={`bi bi-inboxes-fill ${style["empty-inbox"]}`} />
      <p>
        <strong>You have not yet granted access to anyone.</strong>
      </p>
      <p>
        Once you have granted access, you will be able to see it on this page.
      </p>
    </div>
  );
}

function AgentTable({
  agents,
  onDetails,
  searchString,
}: {
  agents: string[];
  onDetails: (_agent: string) => void;
  searchString: string | undefined;
}) {
  const { nameRecord } = useContext(WorkerContext);
  const state = useMemo(
    () =>
      agents
        .map((e) => ({
          key: e,
          webId: e,
          name: nameRecord[e] ?? e,
        }))
        .filter(
          (agent) =>
            !searchString ||
            agent.key.toLowerCase().includes(searchString.toLowerCase()) ||
            agent.name.toLowerCase().includes(searchString.toLowerCase()),
        ),
    [agents, nameRecord, searchString],
  );
  const { orderedData, sortBy, inverted } = useOrdered(headings, state);
  return (
    <div className={styles["table-container"]}>
      <Table className={styles["table-style"]}>
        <Thead headings={headings} inverted={inverted} onSort={sortBy} />
        <tbody data-testid="table-body">
          {orderedData.map(({ key }) => (
            <tr
              key={key}
              className={styles.row}
              onClick={() => {
                onDetails(key);
              }}
              data-testid={`agent-row[${key}]`}
            >
              <td aria-label={`Icon for ${nameRecord[key] ?? key}`}>
                <DisplayImage url={key} fallback />
              </td>
              <td className={styles["table-entry"]} data-testid="agent-name">
                <DisplayAgent url={key} />
              </td>
              <td className={styles["table-entry"]}>{key}</td>
              <td>
                <DetailsIcon />
              </td>
            </tr>
          ))}
        </tbody>
      </Table>
    </div>
  );
}

export default function ListAgents({
  agents,
  onDetails,
  searchString,
}: {
  agents: string[];
  searchString?: string;
  onDetails: (_agent: string) => void;
}) {
  if (agents.length === 0) {
    return <EmptyAgentPage />;
  }
  return (
    <>
      <b data-testid="agent-list-header">
        Organizations, Apps and Companies with Granted Access
      </b>
      <br />
      <br />
      <AgentTable
        agents={agents}
        onDetails={onDetails}
        searchString={searchString}
      />
    </>
  );
}
