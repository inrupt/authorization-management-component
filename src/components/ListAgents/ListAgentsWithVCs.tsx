//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import { Session } from "@inrupt/solid-client-authn-browser";
import { useContext, useEffect, useState } from "react";
import { isValidAccessGrant } from "@inrupt/solid-client-access-grants";
import { VerifiableCredentialApiConfiguration } from "@inrupt/solid-client-vc/dist/common/common";
import getRequestors from "../../helpers/requestors/requestors";
import { SessionContext } from "../../session/SessionProvider";
import Loader from "../Loader/Loader";
import ListAgents from "./ListAgents";
import { WorkerContext } from "../../session/WorkerProvider";

async function getAgents(
  accessEndpoint: string | undefined,
  endpointConfig: VerifiableCredentialApiConfiguration | undefined,
  session: Session,
  _isValidAccessGrant: (
    ..._args: Parameters<typeof isValidAccessGrant>
  ) => Promise<boolean>
) {
  const { webId } = session.info;
  const { fetch } = session;

  if (
    accessEndpoint === undefined ||
    endpointConfig === undefined ||
    typeof webId !== "string"
  ) {
    return [];
  }

  return [
    ...(await getRequestors({
      fetch,
      webId,
      accessEndpoint,
      isValidAccessGrant: _isValidAccessGrant,
      endpointConfiguration: endpointConfig,
    })),
  ];
}

function SuspendedList({
  onDetails,
  searchString,
  agents,
}: {
  onDetails: (_agent: string) => void;
  searchString: string;
  agents?: string[];
}) {
  if (agents === undefined) {
    return <Loader />;
  }
  return (
    <ListAgents
      agents={agents}
      onDetails={onDetails}
      searchString={searchString}
    />
  );
}

export default function ListAgentsWithVCs({
  onDetails,
  searchString,
  onNotEmptyAgents,
}: {
  onDetails: (_agent: string) => void;
  searchString: string;
  onNotEmptyAgents?: () => void;
}) {
  const { session, accessEndpoint, endpointConfiguration } =
    useContext(SessionContext);
  const { isValidAccessGrant } = useContext(WorkerContext);
  const [agents, setAgents] = useState<string[] | undefined>([]);

  useEffect(() => {
    if (accessEndpoint !== undefined && endpointConfiguration !== undefined) {
      // Use undefined value for agent list as a proxy for loading state.
      setAgents(undefined);
      getAgents(
        accessEndpoint,
        endpointConfiguration,
        session,
        isValidAccessGrant
      )
        .then(setAgents)
        .catch((error) => {
          console.error(error);
          setAgents([]);
        });
    }
  }, [accessEndpoint, endpointConfiguration, session, isValidAccessGrant]);

  useEffect(() => {
    if (agents !== undefined && agents.length > 0) {
      if (onNotEmptyAgents) {
        onNotEmptyAgents();
      }
    }
  }, [agents]);

  return (
    <SuspendedList
      agents={agents}
      onDetails={onDetails}
      searchString={searchString}
    />
  );
}
