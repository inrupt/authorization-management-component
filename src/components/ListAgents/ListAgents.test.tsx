// MIT License
//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import { render, waitFor, act } from "@testing-library/react";
import * as axe from "axe-core";
import ListAgents from "./ListAgents";
import WorkerProvider from "../../session/WorkerProvider.mock";

describe("ListAgents", () => {
  it("renders a ListResources component", async () => {
    const onDetails = jest.fn();

    const { asFragment, getByText, getByTestId, getAllByTestId, container } =
      render(
        <WorkerProvider>
          <ListAgents
            agents={[
              "https://id.inrupt.com/jeswrtest51",
              "https://id.inrupt.com/testuser12345?lookup",
            ]}
            onDetails={onDetails}
            onReload={() => {
              /* noop */
            }}
          />
        </WorkerProvider>,
      );

    await waitFor(() => {
      expect(getByText("Jesse Wright 🐨")).toBeVisible();
      expect(getByText("Virginia")).toBeVisible();
    });

    getByText("Jesse Wright 🐨").click();
    getByText("Virginia").click();
    getByText("https://id.inrupt.com/jeswrtest51").click();
    getByText("https://id.inrupt.com/testuser12345?lookup").click();
    expect(onDetails).toHaveBeenCalledTimes(4);
    expect(onDetails).toHaveBeenNthCalledWith(
      1,
      "https://id.inrupt.com/jeswrtest51",
    );
    expect(onDetails).toHaveBeenNthCalledWith(
      2,
      "https://id.inrupt.com/testuser12345?lookup",
    );
    expect(onDetails).toHaveBeenNthCalledWith(
      3,
      "https://id.inrupt.com/jeswrtest51",
    );
    expect(onDetails).toHaveBeenNthCalledWith(
      4,
      "https://id.inrupt.com/testuser12345?lookup",
    );

    // Clicking on the row *should* trigger onDetails
    getByTestId("agent-row[https://id.inrupt.com/jeswrtest51]").click();
    expect(onDetails).toHaveBeenNthCalledWith(
      5,
      "https://id.inrupt.com/jeswrtest51",
    );

    getByTestId(
      "agent-row[https://id.inrupt.com/testuser12345?lookup]",
    ).click();
    expect(onDetails).toHaveBeenNthCalledWith(
      6,
      "https://id.inrupt.com/testuser12345?lookup",
    );

    expect(asFragment()).toMatchSnapshot();

    // Clicking the sort icon for name should toggle the icon
    expect(getByTestId("table-heading-sort-icon-name")).toHaveClass(
      "bi bi-sort-down sort-icon",
    );

    expect(
      getAllByTestId("agent-row", { exact: false }).map((elem) =>
        elem.getAttribute("data-testid"),
      ),
    ).toEqual([
      "agent-row[https://id.inrupt.com/jeswrtest51]",
      "agent-row[https://id.inrupt.com/testuser12345?lookup]",
    ]);

    act(() => {
      getByTestId("table-heading-sort-icon-name").click();
    });

    expect(getByTestId("table-heading-sort-icon-name")).toHaveClass(
      "bi bi-sort-up sort-icon",
    );

    expect(
      getAllByTestId("agent-row", { exact: false }).map((elem) =>
        elem.getAttribute("data-testid"),
      ),
    ).toEqual([
      "agent-row[https://id.inrupt.com/testuser12345?lookup]",
      "agent-row[https://id.inrupt.com/jeswrtest51]",
    ]);

    // WebId should only invert after the second click since it is not in focus
    expect(getByTestId("table-heading-sort-icon-webId")).toHaveClass(
      "bi bi-sort-down sort-icon",
    );

    act(() => {
      getByTestId("table-heading-sort-icon-webId").click();
    });

    expect(getByTestId("table-heading-sort-icon-webId")).toHaveClass(
      "bi bi-sort-down sort-icon",
    );

    expect(
      getAllByTestId("agent-row", { exact: false }).map((elem) =>
        elem.getAttribute("data-testid"),
      ),
    ).toEqual([
      "agent-row[https://id.inrupt.com/jeswrtest51]",
      "agent-row[https://id.inrupt.com/testuser12345?lookup]",
    ]);

    act(() => {
      getByTestId("table-heading-sort-icon-webId").click();
    });

    expect(getByTestId("table-heading-sort-icon-webId")).toHaveClass(
      "bi bi-sort-up sort-icon",
    );

    expect(
      getAllByTestId("agent-row", { exact: false }).map((elem) =>
        elem.getAttribute("data-testid"),
      ),
    ).toEqual([
      "agent-row[https://id.inrupt.com/testuser12345?lookup]",
      "agent-row[https://id.inrupt.com/jeswrtest51]",
    ]);

    // Accessibility tests
    HTMLCanvasElement.prototype.getContext = jest.fn();
    const results = await axe.run(container);
    expect(results.violations).toHaveLength(0);
  });

  it("displays a modal when there are no agents", async () => {
    const onDetails = jest.fn();

    const { getByText, container, baseElement } = render(
      <WorkerProvider>
        <ListAgents
          agents={[]}
          onDetails={onDetails}
          onReload={() => {
            /* noop */
          }}
        />
      </WorkerProvider>,
    );

    await waitFor(() => {
      expect(
        getByText("You have not yet granted access to anyone."),
      ).toBeVisible();
    });

    expect(baseElement).toMatchSnapshot();

    // Accessibility tests
    HTMLCanvasElement.prototype.getContext = jest.fn();
    const results = await axe.run(container);
    expect(results.violations).toHaveLength(0);
  });

  it("renders a ListResources component with an search value", async () => {
    const onDetails = jest.fn();

    const { getByText, getAllByTestId, container } = render(
      <WorkerProvider>
        <ListAgents
          agents={[
            "https://id.inrupt.com/jeswrtest51",
            "https://id.inrupt.com/testuser12345?lookup",
          ]}
          onDetails={onDetails}
          searchString="jess"
          onReload={() => {
            /* noop */
          }}
        />
      </WorkerProvider>,
    );

    await waitFor(() => {
      expect(getByText("Jesse Wright 🐨")).toBeVisible();
    });

    expect(
      getAllByTestId("agent-row", { exact: false }).map((elem) =>
        elem.getAttribute("data-testid"),
      ),
    ).toEqual(["agent-row[https://id.inrupt.com/jeswrtest51]"]);

    // Accessibility tests
    HTMLCanvasElement.prototype.getContext = jest.fn();
    const results = await axe.run(container);
    expect(results.violations).toHaveLength(0);
  });

  it("renders a ListResources component with a search value in the webid but not name", async () => {
    const onDetails = jest.fn();

    const { getByText, getAllByTestId, container } = render(
      <WorkerProvider>
        <ListAgents
          agents={[
            "https://id.inrupt.com/jeswrtest51",
            "https://id.inrupt.com/testuser12345?lookup",
          ]}
          onDetails={onDetails}
          searchString="jeswrtest51"
          onReload={() => {
            /* noop */
          }}
        />
      </WorkerProvider>,
    );

    await waitFor(() => {
      expect(getByText("Jesse Wright 🐨")).toBeVisible();
    });

    expect(
      getAllByTestId("agent-row", { exact: false }).map((elem) =>
        elem.getAttribute("data-testid"),
      ),
    ).toEqual(["agent-row[https://id.inrupt.com/jeswrtest51]"]);

    // Accessibility tests
    HTMLCanvasElement.prototype.getContext = jest.fn();
    const results = await axe.run(container);
    expect(results.violations).toHaveLength(0);
  });

  it("renders a ListResources component with a search value containing captials not in the result", async () => {
    const onDetails = jest.fn();

    const { getByText, getAllByTestId, container } = render(
      <WorkerProvider>
        <ListAgents
          agents={[
            "https://id.inrupt.com/jeswrtest51",
            "https://id.inrupt.com/testuser12345?lookup",
          ]}
          onDetails={onDetails}
          searchString="JESSE"
          onReload={() => {
            /* noop */
          }}
        />
      </WorkerProvider>,
    );

    await waitFor(() => {
      expect(getByText("Jesse Wright 🐨")).toBeVisible();
    });

    expect(
      getAllByTestId("agent-row", { exact: false }).map((elem) =>
        elem.getAttribute("data-testid"),
      ),
    ).toEqual(["agent-row[https://id.inrupt.com/jeswrtest51]"]);

    // Accessibility tests
    HTMLCanvasElement.prototype.getContext = jest.fn();
    const results = await axe.run(container);
    expect(results.violations).toHaveLength(0);
  });

  it("renders a ListResources component with an emoji search value", async () => {
    const onDetails = jest.fn();

    const { getByText, getAllByTestId, container } = render(
      <WorkerProvider>
        <ListAgents
          agents={[
            "https://id.inrupt.com/jeswrtest51",
            "https://id.inrupt.com/testuser12345?lookup",
          ]}
          onDetails={onDetails}
          searchString="🐨"
          onReload={() => {
            /* noop */
          }}
        />
      </WorkerProvider>,
    );

    await waitFor(() => {
      expect(getByText("Jesse Wright 🐨")).toBeVisible();
    });

    expect(
      getAllByTestId("agent-row", { exact: false }).map((elem) =>
        elem.getAttribute("data-testid"),
      ),
    ).toEqual(["agent-row[https://id.inrupt.com/jeswrtest51]"]);

    // Accessibility tests
    HTMLCanvasElement.prototype.getContext = jest.fn();
    const results = await axe.run(container);
    expect(results.violations).toHaveLength(0);
  });
});
