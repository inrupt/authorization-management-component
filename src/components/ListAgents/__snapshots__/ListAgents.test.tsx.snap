// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListAgents displays a modal when there are no agents 1`] = `
<body>
  <div>
    <div
      class="empty-grant-container"
    >
      <i
        class="bi bi-inboxes-fill empty-inbox"
      />
      <p>
        <strong>
          You have not yet granted access to anyone.
        </strong>
      </p>
      <p>
        Once you have granted access, you will be able to see it on this page.
      </p>
    </div>
  </div>
</body>
`;

exports[`ListAgents renders a ListResources component 1`] = `
<DocumentFragment>
  <b
    data-testid="agent-list-header"
  >
    Organizations, Apps and Companies with Granted Access
  </b>
  <br />
  <br />
  <div
    class="table-container"
  >
    <table
      class="table-style table"
    >
      <thead>
        <tr>
          <th
            data-testid="table-heading-sort-image"
          >
            <span
              class="visually-hidden"
            >
              Icon
            </span>
          </th>
          <th
            data-testid="table-heading-sort-name"
          >
            Name
            <i
              aria-label="Sort by name"
              class="bi bi-sort-down sort-icon"
              data-testid="table-heading-sort-icon-name"
              role="button"
              tabindex="0"
              title="Sort by name"
            />
          </th>
          <th
            data-testid="table-heading-sort-webId"
          >
            WebID
            <i
              aria-label="Sort by webId"
              class="bi bi-sort-down sort-icon"
              data-testid="table-heading-sort-icon-webId"
              role="button"
              tabindex="0"
              title="Sort by webId"
            />
          </th>
          <th
            data-testid="table-heading-sort-key"
          >
            Resources
          </th>
        </tr>
      </thead>
      <tbody
        data-testid="table-body"
      >
        <tr
          class="row"
          data-testid="agent-row[https://id.inrupt.com/jeswrtest51]"
        >
          <td
            aria-label="Icon for Jesse Wright üê®"
          >
            <img
              alt="Icon for Jesse Wright üê®"
              class="icon"
              src="https://avatars.githubusercontent.com/u/63333554?v=4"
              title="Icon for Jesse Wright üê®"
            />
          </td>
          <td
            class="table-entry"
            data-testid="agent-name"
          >
            <span>
              Jesse Wright üê®
            </span>
          </td>
          <td
            class="table-entry"
          >
            https://id.inrupt.com/jeswrtest51
          </td>
          <td>
            <i
              class="bi bi-info-circle details-icon"
              data-testid="details-icon"
              role="button"
              tabindex="0"
              title="Open Resource Details"
            />
          </td>
        </tr>
        <tr
          class="row"
          data-testid="agent-row[https://id.inrupt.com/testuser12345?lookup]"
        >
          <td
            aria-label="Icon for Virginia"
          >
            <img
              alt="Icon for Virginia"
              class="icon"
              src="https://virginia.solidcommunity.net/profile/80e79c10-26b2-11ed-a130-d5bd0f09344f.jpeg"
              title="Icon for Virginia"
            />
          </td>
          <td
            class="table-entry"
            data-testid="agent-name"
          >
            <span>
              Virginia
            </span>
          </td>
          <td
            class="table-entry"
          >
            https://id.inrupt.com/testuser12345?lookup
          </td>
          <td>
            <i
              class="bi bi-info-circle details-icon"
              data-testid="details-icon"
              role="button"
              tabindex="0"
              title="Open Resource Details"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
